From f51fd6b2f2576d5e7fdf13b58a79ac1faa48e133 Mon Sep 17 00:00:00 2001
From: Amith A <quic_amitajit@quicinc.com>
Date: Mon, 14 Oct 2024 16:22:49 +0530
Subject: [PATCH] UPSTREAM:HOSTAPD: Update RegInfo control field in HE_OP IE
 for Indoor SP AP

There may exist LPI clients who cannot connect to Indoor SP AP. To solve
this problem, IEEE standard introduces a concept called composite AP,
where the (composite) Indoor AP can serve the following client types:
LPI only certified, SP only certified and Dual certified.

To achieve this
1) Obtain the ext_feature flag for the support of composite AP and add
   it to the driver capabilities flags.
2) Change value of HE_REG_INFO_6GHZ_AP_TYPE_INDOOR_SP to 0x8
3) Advertise HE_REG_INFO_6GHZ_AP_TYPE_INDOOR_SP in RegInfo control field
   in HE_OP for composite AP.

Tag: 6G_REGULATORY
Patch-dependency: r32-002-hostapd-Add-AFC-power-compliant-TPE-for-Standard-pow.patch

Signed-off-by: Amith A <quic_amitajit@quicinc.com>
---
 hostapd/config_file.c        |  2 ++
 src/ap/ap_config.c           |  1 +
 src/ap/ap_config.h           |  2 ++
 src/ap/ieee802_11.h          |  1 -
 src/ap/ieee802_11_he.c       | 10 ++++++++--
 src/common/ieee802_11_defs.h |  2 +-
 6 files changed, 14 insertions(+), 4 deletions(-)

--- a/hostapd/config_file.c
+++ b/hostapd/config_file.c
@@ -5226,6 +5226,8 @@ static int hostapd_config_fill(struct ho
 		conf->disable_csa_dfs = atoi(pos);
 	} else if (os_strcmp(buf, "use_ru_puncture_dfs") == 0) {
 		conf->use_ru_puncture_dfs = atoi(pos);
+	} else if (os_strcmp(buf, "enable_6ghz_composite_ap") == 0) {
+		conf->enable_6ghz_composite_ap = atoi(pos);
 	} else {
 		wpa_printf(MSG_ERROR,
 			   "Line %d: unknown configuration item '%s'",
--- a/src/ap/ap_config.c
+++ b/src/ap/ap_config.c
@@ -337,6 +337,7 @@ struct hostapd_config * hostapd_config_d
 	conf->airtime_update_interval = AIRTIME_DEFAULT_UPDATE_INTERVAL;
 #endif /* CONFIG_AIRTIME_POLICY */
 	conf->group_size = MULTI_MBSSID_GROUP_SIZE_DEFAULT;
+	conf->enable_6ghz_composite_ap = 1;
 
 	hostapd_set_and_check_bw320_offset(conf, 0);
 
--- a/src/ap/ap_config.h
+++ b/src/ap/ap_config.h
@@ -1306,6 +1306,12 @@ struct hostapd_config {
 #define MULTI_MBSSID_GROUP_SIZE_MAX 16
 #define MULTI_MBSSID_GROUP_SIZE_DEFAULT 4
 	u8 group_size;
+
+	/* Config to enable composite AP mode.
+	 * Will be enabled by default if not explicitly disabled in the
+	 * hostpd config.
+	 */
+	bool enable_6ghz_composite_ap;
 };
 
 
--- a/src/ap/ieee802_11_he.c
+++ b/src/ap/ieee802_11_he.c
@@ -265,8 +265,14 @@ u8 * hostapd_eid_he_operation(struct hos
 		else
 			control = center_idx_to_bw_6ghz(seg0);
 
-		control |= hapd->iconf->he_6ghz_reg_pwr_type <<
-			HE_6GHZ_OPER_INFO_CTRL_REG_INFO_SHIFT;
+		if (hapd->iconf->he_6ghz_reg_pwr_type == HE_REG_INFO_6GHZ_AP_TYPE_SP &&
+		    hapd->iconf->enable_6ghz_composite_ap) {
+			control |= HE_REG_INFO_6GHZ_AP_TYPE_INDOOR_SP <<
+				HE_6GHZ_OPER_INFO_CTRL_REG_INFO_SHIFT;
+		} else {
+			control |= hapd->iconf->he_6ghz_reg_pwr_type <<
+				HE_6GHZ_OPER_INFO_CTRL_REG_INFO_SHIFT;
+		}
 
 		if (center_idx_to_bw_6ghz(seg0) &&
 		    !is_6ghz_psc_frequency(ieee80211_chan_to_freq(NULL,
--- a/src/common/ieee802_11_defs.h
+++ b/src/common/ieee802_11_defs.h
@@ -2597,7 +2597,7 @@ enum he_reg_info_6ghz_ap_type {
 	HE_REG_INFO_6GHZ_AP_TYPE_SP		= 1,
 	HE_REG_INFO_6GHZ_AP_TYPE_VLP		= 2,
 	HE_REG_INFO_6GHZ_AP_TYPE_INDOOR_ENABLED	= 3,
-	HE_REG_INFO_6GHZ_AP_TYPE_INDOOR_SP	= 4,
+	HE_REG_INFO_6GHZ_AP_TYPE_INDOOR_SP	= 8,
 	HE_REG_INFO_6GHZ_AP_TYPE_MAX = HE_REG_INFO_6GHZ_AP_TYPE_INDOOR_SP,
 };
 
--- a/hostapd/hostapd.conf
+++ b/hostapd/hostapd.conf
@@ -3489,3 +3489,9 @@ own_ip_addr=127.0.0.1
 # when Multi MBSSID group support is enabled for the radio.
 # This value should be in power of 2.
 #mbssid_group_size=4
+
+##### Enable Composite AP mode in 6GHz ########################################
+#
+# 0 = disabled
+# 1 = enabled (default)
+#enable_6ghz_composite_ap=1
